File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        ListNode* curr = head;
        int size = 0; 
        while(curr){
            size++; 
            curr = curr->next; 
        }

        int target = size - n; 
        if(target == 0) return head->next; 
        curr = head;
        while(curr){
            target--; 
            if(target == 0){
                curr->next = curr->next->next; 
                return head; 
            }
            curr = curr->next; 
        }
        return curr; 
    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = dnext; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode slow = head; 
        ListNode fast = head; 
        while(fast != null && n > 0){
            fast = fast.next; 
            n--; 
        }

        while(fast != null && fast.next != null){
            slow = slow.next; 
            fast = fast.next; 
        }

        if(slow == fast) return null; 
        if(fast == null) return head.next; 
        slow.next = slow.next.next; 

        return head; 
    }
}

//[1,2]
//1 -> 2 -> null 
//2 